--- ../bls/ffi/go/bls/bls.go	2019-11-15 11:37:36.191800373 +0900
+++ bls/bls.go	2019-12-07 20:36:59.331495858 +0900
@@ -1,23 +1,19 @@
 package bls
 
 /*
-#cgo bn256 CFLAGS:-DMCLBN_FP_UNIT_SIZE=4
-#cgo bn256 LDFLAGS:-lbls256
-#cgo bn384 CFLAGS:-DMCLBN_FP_UNIT_SIZE=6
-#cgo bn384 LDFLAGS:-lbls384
-#cgo bn384_256 CFLAGS:-DMCLBN_FP_UNIT_SIZE=6 -DMCLBN_FR_UNIT_SIZE=4
-#cgo bn384_256 LDFLAGS:-lbls384_256
-#cgo LDFLAGS:-lcrypto -lgmp -lgmpxx -lstdc++
-
-#cgo bn256_swapg CFLAGS:-DMCLBN_FP_UNIT_SIZE=4 -DBLS_SWAP_G
-#cgo bn256_swapg LDFLAGS:-lbls256
-#cgo bn384_swapg CFLAGS:-DMCLBN_FP_UNIT_SIZE=6 -DBLS_SWAP_G
-#cgo bn384_swapg LDFLAGS:-lbls384
-#cgo bn384_256_swapg CFLAGS:-DMCLBN_FP_UNIT_SIZE=6 -DMCLBN_FR_UNIT_SIZE=4 -DBLS_SWAP_G
-#cgo bn384_256_swapg LDFLAGS:-lbls384_256
-#cgo LDFLAGS:-lcrypto -lgmp -lgmpxx -lstdc++
+#cgo CFLAGS:-I${SRCDIR}./include -DBLS_ETH -DBLS_SWAP_G
+#cgo LDFLAGS:-lbls384_256 -lstdc++ -lm
+#cgo ios LDFLAGS:-L${SRCDIR}/lib/ios
+#cgo android,arm64 LDFLAGS:-L${SRCDIR}/lib/android/arm64-v8a
+#cgo android,arm LDFLAGS:-L${SRCDIR}/lib/android/armeabi-v7a
+#cgo android,amd64 LDFLAGS:-L${SRCDIR}/lib/android/x86_64
+#cgo linux,amd64 LDFLAGS:-L${SRCDIR}/lib/linux/amd64
+#cgo darwin,amd64 LDFLAGS:-L${SRCDIR}/lib/darwin/amd64
+#cgo darwin,arm64 LDFLAGS:-L${SRCDIR}/lib/darwin/arm64
+#cgo windows,amd64 LDFLAGS:-L${SRCDIR}/lib/windows/amd64
 typedef unsigned int (*ReadRandFunc)(void *, void *, unsigned int);
 int wrapReadRandCgo(void *self, void *buf, unsigned int n);
+#include <mcl/bn_c384_256.h>
 #include <bls/bls.h>
 */
 import "C"
@@ -667,6 +663,48 @@ func (sig *Sign) VerifyHashWithDomain(pu
 	return C.blsVerifyHashWithDomain(&sig.v, &pub.v, (*C.uchar)(unsafe.Pointer(&hashWithDomain[0]))) == 1
 }
 
+// SerializeUncompressed --
+func (pub *PublicKey) SerializeUncompressed() []byte {
+	buf := make([]byte, 2048)
+	// #nosec
+	n := C.blsPublicKeySerializeUncompressed(unsafe.Pointer(&buf[0]), C.mclSize(len(buf)), &pub.v)
+	if n == 0 {
+		panic("err blsPublicKeySerializeUncompressed")
+	}
+	return buf[:n]
+}
+
+// SerializeUncompressed --
+func (sig *Sign) SerializeUncompressed() []byte {
+	buf := make([]byte, 2048)
+	// #nosec
+	n := C.blsSignatureSerializeUncompressed(unsafe.Pointer(&buf[0]), C.mclSize(len(buf)), &sig.v)
+	if n == 0 {
+		panic("err blsSignatureSerializeUncompressed")
+	}
+	return buf[:n]
+}
+
+// DeserializeUncompressed --
+func (pub *PublicKey) DeserializeUncompressed(buf []byte) error {
+	// #nosec
+	err := C.blsPublicKeyDeserializeUncompressed(&pub.v, unsafe.Pointer(&buf[0]), C.mclSize(len(buf)))
+	if err == 0 {
+		return fmt.Errorf("err blsPublicKeyDeserializeUncompressed %x", buf)
+	}
+	return nil
+}
+
+// DeserializeUncompressed --
+func (sig *Sign) DeserializeUncompressed(buf []byte) error {
+	// #nosec
+	err := C.blsSignatureDeserializeUncompressed(&sig.v, unsafe.Pointer(&buf[0]), C.mclSize(len(buf)))
+	if err == 0 {
+		return fmt.Errorf("err blsSignatureDeserializeUncompressed %x", buf)
+	}
+	return nil
+}
+
 // VerifyAggregateHashWithDomain --
 // hashWithDomains is array of 40 * len(pubVec)
 func (sig *Sign) VerifyAggregateHashWithDomain(pubVec []PublicKey, hashWithDomains []byte) bool {
@@ -674,13 +712,12 @@ func (sig *Sign) VerifyAggregateHashWith
 		return false
 	}
 	n := len(pubVec)
-	if n == 0 || len(hashWithDomains) != n * 40 {
+	if n == 0 || len(hashWithDomains) != n*40 {
 		return false
 	}
 	return C.blsVerifyAggregatedHashWithDomain(&sig.v, &pubVec[0].v, (*[40]C.uchar)(unsafe.Pointer(&hashWithDomains[0])), C.mclSize(n)) == 1
 }
 
-
 ///
 
 var sRandReader io.Reader
