--- ../bls/ffi/go/bls/mcl.go	2019-10-27 11:58:06.110727404 +0900
+++ bls/mcl.go	2019-11-07 12:23:26.953922539 +0900
@@ -1,10 +1,8 @@
 package bls
 
 /*
-#cgo bn256 CFLAGS:-DMCLBN_FP_UNIT_SIZE=4
-#cgo bn384 CFLAGS:-DMCLBN_FP_UNIT_SIZE=6
-#cgo bn384_256 CFLAGS:-DMCLBN_FP_UNIT_SIZE=6 -DMCLBN_FR_UNIT_SIZE=4
-#include <mcl/bn.h>
+#cgo CFLAGS:-I./include
+#include <mcl/bn_c384_256.h>
 */
 import "C"
 import "fmt"
